pkg/kube/kclient/client_test.go

```diff
@@ -35,6 +42,75 @@ import (
 	"istio.io/istio/pkg/test/util/retry"
 )
 
+func TestSwappingClient(t *testing.T) {
```

Comment:
@hzxuzhonghu check this out - experimenting with a client that abstracts the CRD watching away.

```
If CRD is already ready, create normal client
If CRD is not ready:
  While CRDWatcher not synced, return HasSynced=false and all empty responses.
  Once CRDWatcher syncs:
     If have CRD: swap to a normal client
     Else, mark HasSynced=true, return all empty responses. Once the CRD is ready, swap to a normal client
```

It isn't 100% there in current iteration but pretty close

Reply:
tbh not sure this is a good idea but it was fun

Reply:
No need in our current use cases. We could add it though.

Reply:
I improved it up a bit more (still needs just cleaning up). Basically now the fact that its a "delayed until CRD is ready" is completely transparent.